DDG4/plugins/Geant4TVUserParticleHandler.cpp

```diff
@@ -0,0 +1,160 @@
+     * @version 1.0
+     */
+    class Geant4TVUserParticleHandler : public Geant4UserParticleHandler  {
+      Volume m_trackingVolume;
+
+    public:
+      /// Standard constructor
+      Geant4TVUserParticleHandler(Geant4Context* context, const std::string& nam);
+
+      /// Default destructor
+      virtual ~Geant4TVUserParticleHandler() {}
+
+      /// Post-track action callback
+      /** Allow the user to force the particle handling in the post track action
+       *  set the reason mask to NULL in order to drop the particle.
+       *  The parent's reasoning mask will be or'ed with the particle's mask
+       *  to preserve the MC truth for the hit creation.
+       *  The default implementation is empty.
+       *
+       *  Note: The particle passed is a temporary and will be copied if kept.
+       */
+      virtual void end(const G4Track* track, Particle& particle);
+
+      /// Post-event action callback: avoid warning (...) was hidden [-Woverloaded-virtual]
+      virtual void end(const G4Event* event);
+
+    };
+  }    // End namespace sim
+}      // End namespace dd4hep
+
+#endif // DD4HEP_DDG4_GEANT4TVUSERPARTICLEHANDLER_H
+
+//====================================================================
+//  AIDA Detector description implementation
+//--------------------------------------------------------------------
+//
+//  Author     : M.Frank
+//
+//====================================================================
+// Framework include files
+//#include <DDG4/Geant4TVUserParticleHandler.h>
+#include <DDG4/Geant4Particle.h>
+#include <DDG4/Factories.h>
+#include <DDG4/Geant4Kernel.h>
+
+
+using namespace dd4hep::sim;
+DECLARE_GEANT4ACTION(Geant4TVUserParticleHandler)
+
+/// Standard constructor
+Geant4TVUserParticleHandler::Geant4TVUserParticleHandler(Geant4Context* ctxt, const std::string& nam)
+: Geant4UserParticleHandler(ctxt,nam)
+{
+  m_trackingVolume = ctxt->kernel().detectorDescription().trackingVolume();
+}
+
+/// Post-track action callback
+void Geant4TVUserParticleHandler::end(const G4Track* /* track */, Particle& p)  {
+
+  std::array<double, 3> start_point = {p.vsx, p.vsy, p.vsz};
+  bool starts_in_trk_vol = m_trackingVolume.ptr()->Contains(start_point.data());
+
+  dd4hep::detail::ReferenceBitMask<int> reason(p.reason);
+
+  if( reason.isSet(G4PARTICLE_PRIMARY) ) {
+    //do nothing
+  } else if( starts_in_trk_vol && ! reason.isSet(G4PARTICLE_ABOVE_ENERGY_THRESHOLD) )  {
+    // created in tracking volume but below energy cut
+    p.reason = 0;
+    return;
+  }
+
+  std::array<double, 3> end_point = {p.vex, p.vey, p.vez};
+  bool ends_in_trk_vol = m_trackingVolume.ptr()->Contains(end_point.data());
+
+  // created and ended in calo but not primary particle
+  //
+  // we can have particles from the generator only in the calo, if we have a
+  // long particle with preassigned decay, we need to keep the reason or the
+  // MChistory will not be updated later on
+  if( not reason.isSet(G4PARTICLE_PRIMARY) ) {
```

Comment:
I wonder if we can share this code between the TC and TV handlers?

Reply:
that would be the clean way to do it yes :(